File: 016-retirement_py/retirement.py
def retirement(start_age, initial_savings, working_info, retired_info):
    """
    Prints the current status of an individual's retirement account.
    The dictionaries both have these fields: 
    "months","contribution","rate_of_return"

    Args:
        start_age (int): At what age (in months) does the individual start
        initial_savings (float): initial savings in dollars
        working_info (dict): information about working
        retired_into (dict): information about retirement

Returns:
None
    """
    
    working_balance = calculate_balance(
        working_info['months'],
        initial_savings,
        working_info['contribution'],
        working_info['rate_of_return'],
        start_age
    )
    
    calculate_balance(
        retired_info['months'],
        working_balance,
        retired_info['contribution'],
        retired_info['rate_of_return'],
        start_age + working_info['months']
    )
    
    
def calculate_balance(months, initial_balance, contribution, rate_of_return, start_age):
    balance = initial_balance
    for month in range(months):
        age_years = (start_age + month) // 12
        age_months = (start_age + month) % 12
        print(f"Age {age_years:3d} month {age_months:2d} you have ${balance:,.2f}")
        balance += balance * rate_of_return + contribution
    return balance

# Working:
#   --------
#    Months: 489
#    Per Month Savings: $1000
#    Rate of Return:  4.5% per year ( 0.045/12 per month) 
#                     [above inflation]
#  Retired:
#  --------
#    Months: 384
#    Per Month Spending: -4000
#    Rate of Return: 1% per year ( 0.01/12 per month) 
#                    [above inflation]
#  Starting conditions:
#  -------------------
#     Age: 327 months (27 years, 3 months)
#     Savings: $21,345


working_info = {
    "months": 489,
    "contribution": 1000,
    "rate_of_return": 0.045 / 12
}

retired_info = {
    "months": 384,
    "contribution": -4000,
    "rate_of_return": 0.01 / 12
}

start_age = 327  
initial_savings = 21345

retirement(start_age, initial_savings, working_info, retired_info)

File: 037-retirement_cpp/retirement.cpp
#include <stdio.h>
#include <stdlib.h>

struct _retire_info {
    int months;
    double contribution;
    double rate_of_return;
};
typedef struct _retire_info retire_info;

double countBalance(double balance, retire_info a) {
    balance = balance * (1 + a.rate_of_return) + a.contribution;
    return balance;
}

void retirement(int startAge, double initial, retire_info working, retire_info retired) {
    double balance = initial;
    
    // Working phase
    for (int i = 0; i < working.months + 1; i++) {
        int age = (startAge + i) / 12;
        int month = (startAge + i) % 12;
        printf("Age %3d month %2d you have $%.2f\n", age, month, balance);
        if (i == working.months) {
            break;
        }
        balance = countBalance(balance, working);
    }

    // Retired phase - No change in logic here
    for (int j = 1; j < retired.months; j++) {
        int age = (startAge + working.months + j) / 12;
        int month = (startAge + working.months + j) % 12;
        balance = countBalance(balance, retired);
        printf("Age %3d month %2d you have $%.2f\n", age, month, balance);
    }
}

int main(void) {
    int startAge = 327;  // 27 years and 3 months
    double initial = 21345;  // Initial savings of $21,345

    // Working phase: 40 years and 9 months, contributing $1000 per month with 4.5% annual return
    retire_info working = {
        489, 1000.0, 0.045 / 12
    };

    // Retired phase: 32 years, spending $4000 per month with 1% annual return
    retire_info retired = {
        384, -4000.0, 0.01 / 12
    };

    // Perform retirement calculation
    retirement(startAge, initial, working, retired);
    
    return 0;
}
